src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11251,14 +11251,43 @@ private bool ScanCast(bool forPattern = false)
                     // following a cast.
                     return !forPattern || this.CurrentToken.Kind switch
                     {
-                        SyntaxKind.PlusToken => true,
-                        SyntaxKind.MinusToken => true,
-                        SyntaxKind.AmpersandToken => true,
-                        SyntaxKind.AsteriskToken => true,
+                        SyntaxKind.PlusToken or
+                        SyntaxKind.MinusToken or
+                        SyntaxKind.AmpersandToken or
+                        SyntaxKind.AsteriskToken or
                         SyntaxKind.DotDotToken => true,
-                        _ => CanFollowCast(this.CurrentToken.Kind)
+                        SyntaxKind.IdentifierToken => !isBinaryPattern(),
+                        var tk => CanFollowCast(tk)
                     };
 
+                    bool isBinaryPattern()
+                    {
+                        if (!isBinaryPatternKeyword())
+                        {
+                            return false;
+                        }
+
+                        bool lastTokenIsBinaryOperator = true;
+
+                        EatToken();
+                        while (isBinaryPatternKeyword())
+                        {
+                            // If the parenthesized type is followed by an even-number of binary pattern tokens, it is a cast.
+                            // Such as `(int)or or string`
+                            // If the parenthesized type is followed by an odd-number of binary pattern token, it is not a cast.
+                            // Such as `(int) or string`
```

Comment:
This will only affect cases where it's currently an error and also accept the rest as it always has.